Web Inspector: CSS source editor can adopt fuzzy search for code completion

the-chenergy opened the pull request:
#### e058a9c99c8c35dac22dc6095343a4895a6fb168
<pre>
Web Inspector: CSS source editor can adopt fuzzy search for code completion
<a href="https://rdar.apple.com/125030691">rdar://125030691</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=271260">https://bugs.webkit.org/show_bug.cgi?id=271260</a>

Reviewed by Devin Rousso.

Use CSSCompletions.executeQuery instead of CSSCompletions.startsWith
when language mode is CSS, since executeQuery uses fuzzy search.

Also add an setting item in the Settings tab&apos;s Experimental sub-tab to
optionally turn this feature off.

* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
(WI.CodeMirrorCompletionController.prototype._generateCSSCompletions):
   - Optionally apply fuzzy search by using `executeQuery`.

(WI.CodeMirrorCompletionController.prototype.updateCompletions):
   - The argument `completions` could now also be an array of objects
     instead of always an array of strings (due to executeQuery
     returning an array of objects). Those objects also contain info on
     which characters match in the fuzzy search, so the suggestion box
     can bold the matching text. `this._completions` however does not
     need this extra info, so we extract only the texts from
     `completions` for that.

(WI.CodeMirrorCompletionController.prototype._applyCompletionHint):
(WI.CodeMirrorCompletionController.prototype._applyCompletionHint.update):
   - Fuzzy search can now match completions not necessarily starting
     with the same text as currentText. Only show the gray watermark
     text (hint marker in code) when it&apos;s actually a prefix match.

*       Source/WebInspectorUI/UserInterface/Models/Completions.js: Added.
* Source/WebInspectorUI/UserInterface/Models/CSSCompletions.js:
(WI.CSSCompletions.getCompletionText):
   - As the getCompletionText helper function doesn&apos;t have to work for
     only CSS suggestions, it better belongs to a class with a more
     generic name like Completion instead of CSSCompletion.
   - Leave a note to eventually migrate all other usages of the old
     CSSCompletion.getCompletionText to Completion.getCompletionText.

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
   - Include the new Completions.js source file.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
   - Add an experimental setting item for this feature change.

Canonical link: <a href="https://commits.webkit.org/279144@main">https://commits.webkit.org/279144@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc1b9c9f7bad0fcce44fc94a8edebcc4a47b4691

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51052 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30354 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3380 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54309 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1741 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36641 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1412 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41571 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/979 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53151 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27994 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43988 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22691 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25316 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1243 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47296 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1316 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55904 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26160 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1211 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48974 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48101 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11493 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28289 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27141 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -93,24 +93,29 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
 
-        this._completions = completions;
+        let completionTexts;
+        if (completions.length && typeof completions[0] !== "string")
```
we already know that `completions.length` is true based on the above `if`, so you can remove this
```js
this._completions = completions;
if (typeof this._completions[0] !== "string")
    this._completions = this._completions.map(WI.CSSCompletions.getCompletionText);
```

rcaliman-apple replied:
Let's make an explicit check for  `completion[0] instanceof WI.QueryResult` instead of `typeof completions[0] !== "string"`

There is an assertion for this in `CSSCompletions.getCompletionText()` but it helps a future reader know what to expect when they get here, particularly because `WI.CodeMirrorCompletionController` has to keep compatibility with both strings and `WI.QueryResult`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -93,24 +93,29 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
 
         // Try to restore the previous selected index, otherwise just select the first.
-        var index = this._currentCompletion ? completions.indexOf(this._currentCompletion) : 0;
+        let index = this._currentCompletion ? completionTexts.indexOf(this._currentCompletion) : 0;
```
we also should be using the member `this._completions` in the places that `completions` (or `completionsTexts`) is used so that there's only one source of truth
```js
let index = this._currentCompletion ? this._completions.indexOf(this._currentCompletion) : 0;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -93,24 +93,29 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
             index = 0;
 
-        if (this._forced || completions.length > 1 || completions[index] !== this._prefix) {
+        if (this._forced || completions.length > 1 || completionTexts[index] !== this._prefix) {
```
ditto
```js
if (this._forced || this._completions.length > 1 || this._completions[index] !== this._prefix) {
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -125,7 +130,7 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
         }
 
-        this._applyCompletionHint(completions[index]);
+        this._applyCompletionHint(completionTexts[index]);
```
ditto
```js
this._applyCompletionHint(this._completions[index]);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -584,7 +590,7 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
+            let functionCompletions = WI.CSSKeywordCompletions.forFunction(functionName).executeQuery(this._prefix);
 
             if (this._delegate && this._delegate.completionControllerCSSFunctionValuesNeeded)
                 functionCompletions = this._delegate.completionControllerCSSFunctionValuesNeeded(this, functionName, functionCompletions);
```
Aside: nobody seems to use this anymore, so i think we actually could/should just remove it

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -93,24 +93,29 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
-        var bounds = new WI.Rect(firstCharCoords.left, firstCharCoords.top, lastCharCoords.right - firstCharCoords.left, firstCharCoords.bottom - firstCharCoords.top);
+        let firstCharCoords = this._codeMirror.cursorCoords(from);
+        let lastCharCoords = this._codeMirror.cursorCoords(to);
+        let bounds = new WI.Rect(firstCharCoords.left, firstCharCoords.top, lastCharCoords.right - firstCharCoords.left, firstCharCoords.bottom - firstCharCoords.top);
```
NIT: we dont normally make changes like this unless you're actually modifying the line in some other way, so id prefer if we left these as-is in order to make things like `git blame` easier to traverse

dcrousso reviewed (commented):
this kinda sorta feels like something that should be controlled by a `WI.Setting`, similar to how the global search ⇧⌘F has an option to decide whether the text is a regex or exact match 🤔

tho the Styles panel in the details sidebar of the Elements Tab already uses fuzzy match completions, so i guess there's already some precedent and it's not entirely a stretch to also do that in the Sources Tab 😅

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -629,7 +635,7 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
         this._implicitSuffix = suffix !== ":" ? ": " : "";
 
         // Complete property names.
-        return WI.cssManager.propertyNameCompletions.startsWith(this._prefix);
```
I believe this was the last caller for `WI.CSSPropertyNameCompletions.startsWith()` so you can remove it now.

ditto for `WI.CSSCompletions.startsWith()`

the-chenergy replied:
Not the last caller anymore now that we added an experimental setting for this 😦

rcaliman-apple reviewed (commented):
Looking great! Thank you for bringing this over to the Sources tab 👍

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -94,6 +94,8 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
 
         this._completions = completions;
+        if (completions[0] instanceof WI.QueryResult)
+            this._completions = completions.map(WI.CSSCompletions.getCompletionText);
```
actually, why not just always do this since it already has nice behavior when everything is a string?
```js
this._completions = completions.map(WI.CSSCompletions.getCompletionText);
```

also, i think we should move `getCompletionText` to a more generic class (e.g. `WI.Completions`) since `WI.CodeMirrorCompletionController.prototype.updateCompletions` is technically not specific to any language (i.e. it's sorta a layering violation

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -584,10 +587,7 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
-
-            if (this._delegate && this._delegate.completionControllerCSSFunctionValuesNeeded)
-                functionCompletions = this._delegate.completionControllerCSSFunctionValuesNeeded(this, functionName, functionCompletions);
+            let functionCompletions = WI.CSSKeywordCompletions.forFunction(functionName).executeQuery(this._prefix);
```
we can get rid of this variable and just directly return the value instead
```js
return WI.CSSKeywordCompletions.forFunction(functionName).executeQuery(this._prefix);
```

dcrousso reviewed (commented):
> The advantage of defaulting to the previous selection is that if you highlight some other item and keep typing text matching the selection, it won't suddenly jump to the top if there are other suggestions:
> 
> Which I think... rarely even happens? 😄 So I think changing it to defaulting to the first item instead is also fine.

yeah i suppose it's probably not common for someone to bother changing what completion is highlighted _without_ committing it in order to then type that completion themselves, so i could see an argument for just always resetting back to the first completion (or the most common one if something like `WI.settings.experimentalCSSSortPropertyNameAutocompletionByUsage` is enabled)

again though id want that to be behind a `WI.ExperimentalSetting` (which i'd again be fine having as on-by-default) just in case some really weird edgecase behavior happens which we need to address.  i dont think it's likely, but it's extremely cheap to add a `WI.ExperimentalSetting` so id err on the side of caution just in case :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
```diff
@@ -629,7 +632,10 @@ WI.CodeMirrorCompletionController = class CodeMirrorCompletionController extends
+        if (WI.settings.experimentalUseFuzzyMatchingForCSSCodeCompletion.value)
+            return WI.cssManager.propertyNameCompletions.executeQuery(this._prefix);
+        else
+            return WI.cssManager.propertyNameCompletions.startsWith(this._prefix);
```
Style: you dont need the `else` here as it's already implicit because in order to reach this we must not've done the `if` since it would've returned
```js
if (WI.settings.experimentalUseFuzzyMatchingForCSSCodeCompletion.value)
    return WI.cssManager.propertyNameCompletions.executeQuery(this._prefix);
return WI.cssManager.propertyNameCompletions.startsWith(this._prefix);
```

dcrousso reviewed (approved):
great work! :)

the-chenergy committed:
Web Inspector: CSS source editor can adopt fuzzy search for code completion
rdar://125030691
https://bugs.webkit.org/show_bug.cgi?id=271260

Reviewed by Devin Rousso.

Use CSSCompletions.executeQuery instead of CSSCompletions.startsWith
when language mode is CSS, since executeQuery uses fuzzy search.

Also add an setting item in the Settings tab's Experimental sub-tab to
optionally turn this feature off.

* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorCompletionController.js:
(WI.CodeMirrorCompletionController.prototype._generateCSSCompletions):
   - Optionally apply fuzzy search by using `executeQuery`.

(WI.CodeMirrorCompletionController.prototype.updateCompletions):
   - The argument `completions` could now also be an array of objects
     instead of always an array of strings (due to executeQuery
     returning an array of objects). Those objects also contain info on
     which characters match in the fuzzy search, so the suggestion box
     can bold the matching text. `this._completions` however does not
     need this extra info, so we extract only the texts from
     `completions` for that.

(WI.CodeMirrorCompletionController.prototype._applyCompletionHint):
(WI.CodeMirrorCompletionController.prototype._applyCompletionHint.update):
   - Fuzzy search can now match completions not necessarily starting
     with the same text as currentText. Only show the gray watermark
     text (hint marker in code) when it's actually a prefix match.

*	Source/WebInspectorUI/UserInterface/Models/Completions.js: Added.
* Source/WebInspectorUI/UserInterface/Models/CSSCompletions.js:
(WI.CSSCompletions.getCompletionText):
   - As the getCompletionText helper function doesn't have to work for
     only CSS suggestions, it better belongs to a class with a more
     generic name like Completion instead of CSSCompletion.
   - Leave a note to eventually migrate all other usages of the old
     CSSCompletion.getCompletionText to Completion.getCompletionText.

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
   - Include the new Completions.js source file.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
   - Add an experimental setting item for this feature change.

Canonical link: https://commits.webkit.org/279144@main
